### PR TITLE
Opam 2.0 development version & libraries

### PR DESCRIPTION
**Pull request by AltGr**
Not sure if this should be merged, or at least not right now, but I wanted to open the discussion.

We have a repository https://github.com/OcamlPro/opam-dev-repo which
contains basically the same packages (just not bound to the alpha4
release), so we can already prompt users to use that if they want to
bootstrap opam 2 from opam 1.2.

This can be particularly useful for updating packages depending on
opam-lib. There is no conflict since the package names (and library
names) are all different -- although, that could lead to confusion.

The `opam-devel` package installs opam, opam-admin, opam-installer
(based on the separately packaged libs) to its `lib/opam-devel`
directory and displays a message showing the location and suggesting
to copy them to PATH, outside of the opam installation.

### CODE REVIEW

**Comment by avsm** [comment]
Makes sense to this merge this to me...

**Comment by AltGr** [comment]
Ok, then I'll make the "preview release" announcement reflect that, and wait for it.

**Comment by AltGr** [comment]
@samoht just waiting for your opinion on this ?

**Comment by AltGr** [comment]
About the libraries the separate packages depend on:
- `opam-core`, `opam-format`, `opam-repository`, `opam-state` only require `re`, `jsonm`, `ocamlgraph`
- `opam-solver` additionally requires `dose` (which in turn pulls in `cudf` and `extlib`)
- `opam-client`, `opam-devel` additionally require `cmdliner`.

So if this is too many packages, it would be possible do reduce the granularity and have only 2/3 packages (one or two with minimal dependencies, excluding the solver and actions, and a complete one) instead of 7. But I guess `opkg` likes it better this way ?

**Comment by dbuenzli** [comment]
> But I guess opkg likes it better this way ?

I'm not sure. I think infrastructure tools (i.e. people wanting to read/update OPAM files and OPAM install files) only care about `opam-format`. Why do you need `jsonm`, `re` and `ocamlgraph` for it ? This should have no dependency. The rest I have no use. So maybe you should reduce the granularity (also for your own sanity).

**Comment by dbuenzli** [comment]
> This should have no dependency.

(For bootstrap reasons)

**Comment by AltGr** [comment]
well since these are opam packages, they require a working opam installation, so I don't see the dependencies as a problem: we have other ways already in place to handle the bootstrap without a working opam (inclusive tar.gz, `make lib-ext`, `make cold`...) ; this is mostly a nice way for ppl with opam 1.2 to get opam 2.0.

On the granularity: `opam-format` reads individual files, but `opam-state` has knowledge of the standard location of files below the opam root, and the mechanisms to read the subsystem states from there, so depending on their use, the tools may need that as well.

`jsonm`, `re` and `ocamlgraph` are needed already by `opam-core`, so changing the granularity wouldn't help; `re` is directly used in `OpamStd`, `ocamlgraph` is needed for the parallel command processing engine which is part of `opam-core`, and `jsonm` is useful to define related functions in our base `Set` and `Map` modules. It would be a lot of work to try and remove these dependencies, not just a change in granularity, and they don't seem very prolematic to me. The three of them together don't weight that much.

**Comment by dbuenzli** [comment]
> well since these are opam packages, they require a working opam installation,

People working on tools to convert opam packages to system package managers may not want this.

I don't understand why I need a parallel command processing engine to read opam file formats. I'm not saying you should remove the deps you mention of `opam-core`, I'm saying `opam-core` should depend on `opam-format` which should be distributed as a separate, dependency less, library. We want to be able to use the OPAM file formats outside of opam itself and its machinery.

**Comment by dbuenzli** [comment]
> we have other ways already in place to handle the bootstrap without a working opam

Just to make that clear, it's not about bootstrapping opam it's about general eco-system bootstrapping. `opkg` re-uses the OPAM file format as a meta-data file format. Since it doesn't want to implement its own parser it uses `opam-format`. In the future `jsonm` may use `opkg` to find about its build dependencies, namely `uutf`. With the current structure any dep of `opam-core` cannot use `opkg` to build as it leads to a dep cycle.

**Comment by AltGr** [comment]
Hm, I understand your concern, but it can't be fixed so easily... opam-format uses `opam-core.OpamStd` a lot, and includes high-level i/o functions that do e.g. proper locking of the files to prevent concurrent writes, and that relies on `opam-core.OpamSystem` too. Also, the types returned make use of the base types in turn using the modules of `opam-core` and their json serialisation functions.

In short, OpamParallel and the ocamlgraph dependency could be lifted, but otherwise, a large refactoring would be needed; not sure it is worth it, and at this stage, that wouldn't be before 3.0 anyway.

**Comment by dbuenzli** [comment]
> Hm, I understand your concern, but it can't be fixed so easily... opam-format uses opam-core.OpamStd a lot, and includes high-level i/o functions that do e.g. proper locking of the files to prevent concurrent writes, and that relies on opam-core.OpamSystem too.

My god why do you need all that crap to write file codecs ?  It'a bit ridiculous. Just give me things that are able to codec over channels or a string. The desire has been expressed many times that we would like to have a package only for the file codecs (if only to get a lean `opam-installer`), the fact that we are not getting this in OPAM v2 is quite disappointing; the fact that you think it's not worth it is short-sighted regarding the build future of the eco-system.

**Comment by avsm** [comment]
> the fact that you think it's not worth it is short-sighted regarding the build future of the eco-system.

This is a rather badly targeted sentence.  @AltGr explained the situation quite clearly: the codec codebase already uses the various `opam-core` libraries (and mostly always has).  Extricating it into a separate, dependency-free library is a laudable goal, but one that takes a lot of refactoring work and is a _new feature_ as a result.

So given where we are now in the release cycle, the request has to be judged alongside all the other new feature requests and prioritised accordingly.  Please open a bug report about a dependency free format library.   It also seems quite possible to build an independent implementation outside of the current 2.0 release cycle, if that is an easier solution... they can all be tested alongside each other.

**Comment by dbuenzli** [comment]
> This is a rather badly targeted sentence.

It's not. He says he doesn't think it's worth it. So I'm not going to loose my time again trying to convince. If you think it's worthwhile I let you open a bug yourself. 

Besides the whole codebase was reshuffelled for v2 and it strikes me as odd that what I'm asking didn't naturally emerge out of this. Especially since the desire for this was expressed more than once.

**Comment by AltGr** [comment]
I still think this is worth discussing; but may we please focus on the PR at hand ?

**Comment by avsm** [comment]
@dbuenzli:
> It's not. He says he doesn't think it's worth it.

No, @AltGr actually said:
> not sure it is worth it, and at this stage, that wouldn't be before 3.0 anyway

Since it's a large code shuffle, that would derail the OPAM 2.0 release. It could happen in a future release, or we presumably delay OPAM 2 by quite a bit to accommodate this feature. Your desire can still be accommodated by my other suggestion that a parser library be built independently and later reintegrated into OPAM.

> So I'm not going to loose my time again trying to convince. If you think it's worthwhile I let you open a bug yourself. <...> Especially since the desire for this was expressed more than once.

A pity, since you would have saved your own time by opening a tracking issue with the desire in the first place. I've opened ocaml/opam#2677 to track it now.

File: packages/opam-client/opam-client.2.0~alpha4/descr [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Opam 2.0 development libraries
+
+Actions on the opam root, switches, installations, and front-end.
```
Comment (avsm) [inline packages/opam-client/opam-client.2.0~alpha4/descr]: Opam, opam or OPAM? :-)

File: packages/opam-client/opam-client.2.0~alpha4/opam [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "Guillem Rieu        <guillem.rieu@ocamlpro.com>"
+  "Vincent Bernardoff  <vb@luminar.eu.org>"
+  "Roberto Di Cosmo    <roberto@dicosmo.org>"
+]
+homepage: "https://opam.ocaml.org/"
+bug-reports: "https://github.com/ocaml/opam/issues"
+dev-repo: "https://github.com/ocaml/opam.git"
+build: [
+  ["./configure" "--disable-checks" "--prefix" prefix]
+  [make name]
```
Comment (avsm) [inline packages/opam-client/opam-client.2.0~alpha4/opam]: does this unquoted one work on OPAM 1.2.0?

File: packages/opam-core/opam-core.2.0~alpha4/opam [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+dev-repo: "https://github.com/ocaml/opam.git"
+build: [
+  ["./configure" "--disable-checks" "--prefix" prefix]
+  [make name]
+  [make "%{name}%.install"]
+]
+depends: [
+  "ocamlfind" {build}
+  "ocp-build" {build & >= "1.99.7"}
+  "ocamlgraph"
```
Comment (avsm) [inline packages/opam-core/opam-core.2.0~alpha4/opam]: Isn't there a minimum version of ocamlgraph as well? At least we should put a minimum version based on whats tested in src_ext perhaps.

File: packages/opam-devel/opam-devel.2.0~alpha4/descr [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Opam 2.0 development version
+
+This package compiles the a development version of opam 2.0. For consistency and
```
Comment (avsm) [inline packages/opam-devel/opam-devel.2.0~alpha4/descr]: s/the a/the/g

File: packages/opam-repository/opam-repository.2.0~alpha4/descr [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Opam 2.0 development libraries
+
+This library includes repository and remote sources handling, including
+Curl/wget, rsync, git, mercurial, darcs backends.
```
Comment (avsm) [inline packages/opam-repository/opam-repository.2.0~alpha4/descr]: nit: s/Curl/curl

File: packages/opam-client/opam-client.2.0~alpha4/opam (hunk 2 above)
Comment (AltGr) [inline packages/opam-client/opam-client.2.0~alpha4/opam]: yes, variable `name` is defined in 1.2.0 too; I have tested these packages against 1.2.1 and 1.2.2 (and 1.2.0, but there were some issues).

File: packages/opam-core/opam-core.2.0~alpha4/opam (hunk 3 above)
Comment (AltGr) [inline packages/opam-core/opam-core.2.0~alpha4/opam]: hmm, yes indeed, we use 1.8.5 in lib-ext ; dose requires `1.8.5` already anyway.

**Comment by AltGr** [comment]
All fixed, pls merge! :)
